src/highdicom/seg/sop.py

```diff
@@ -1916,81 +1984,140 @@ def _get_pixels_by_seg_frame(
-                raise RuntimeError(
-                    'Segments cannot be combined because they overlap'
-                )
+            if pixel_array.ndim == 2:
+                pixel_array = pixel_array[None, :, :]
 
-            # Scale the array by the segment number using broadcasting
-            if relabel:
-                pixel_value_map = np.arange(1, len(segment_numbers) + 1)
-            else:
-                pixel_value_map = segment_numbers
-            scaled_array = pixel_array * pixel_value_map.reshape(1, 1, 1, -1)
+            # Initialize empty pixel array
+            out_array = np.zeros((nfo, h, w), dtype=intermediate_dtype)
 
-            # Combine segments by taking maximum down final dimension
-            max_array = scaled_array.max(axis=-1)
-            pixel_array = max_array
+            # Loop over output frames
+            for fo in range(nfo):
+                # Loop over segmentation indices
+                for seg_ind, seg_n in enumerate(segment_numbers):
+                    fi = seg_frames_matrix[fo, seg_ind] - 1
+                    pix_value = intermediate_dtype(
+                        seg_ind + 1 if relabel else seg_n
+                    )
+                    if fi >= 0:
+                        if seg_ind > 0 and not skip_overlap_checks:
+                            if np.any(
+                                np.logical_and(
+                                    pixel_array[fi, :, :] > 0,
+                                    out_array[fo, :, :] > 0
+                                )
+                            ):
+                                raise RuntimeError(
+                                    "Cannot combine segments because segments "
+                                    "overlap."
+                                )
+                        out_array[fo, :, :] = np.maximum(
+                            pixel_array[fi, :, :] * pix_value,
+                            out_array[fo, :, :]
+                        )
 
-        return pixel_array
+        else:
+            # Initialize empty pixel array
+            out_array = np.zeros((nfo, h, w, nseg), intermediate_dtype)
+
+            # Loop through output frames
+            for out_frm, frm_row in enumerate(seg_frames_matrix):
+                # Loop through segmentation frames
+                for out_seg, seg_frame_num in enumerate(frm_row):
+                    if seg_frame_num >= 1:
+                        seg_frame_ind = seg_frame_num.item() - 1
+                        # Copy data to to output array
+                        if self.pixel_array.ndim == 2:
+                            # Special case with a single segmentation frame
+                            out_array[out_frm, :, :, out_seg] = \
+                                self.pixel_array
+                        else:
+                            out_array[out_frm, :, :, out_seg] = \
+                                self.pixel_array[seg_frame_ind, :, :]
```

Comment:
Does this repeatedly decode the pixel data? It may be advantageous to store `pixel_array` as a separate variable outside of the loop instead of repeatedly accessing the attribute/property on the `self` object.

Reply:
No, this is cached on the Dataset at the pydicom level

Reply:
Good to know. I would nevertheless suggest avoiding the repeated attribute access, since it may negatively impact performance. See https://github.com/pydicom/pydicom/issues/1728#issuecomment-1312833040

Reply:
In context of the above comment (https://github.com/herrmannlab/highdicom/pull/208#discussion_r1042309582), we may even want to create a deep copy outside of the loop.

Reply:
A deepcopy out of the loop would be hugely inefficient and the comment you point to doesn't apply here since the `.pixel_array` exists as a Python-level property rather than the loopkup being deferred to pydicom.Dataset's dictionary lookup process